File: 2370-longest-ideal-subsequence/2370-longest-ideal-subsequence.java
class Solution {
    int helper_function(String s,int k,int idx,char ch,int dp[][]){
        if(idx==s.length()) return 0;
        if(dp[idx][ch]!=-1) return dp[idx][ch];
        int pick=0;
             if(Math.abs(ch-s.charAt(idx))<=k || ch=='#') pick=1+helper_function(s,k,idx+1,s.charAt(idx),dp); 
             int notpick=helper_function(s,k,idx+1,ch,dp);
       return dp[idx][ch]=Math.max(pick,notpick);
        
    }
    public int longestIdealString(String s, int k) {
        int n=s.length();
        int dp[][]=new int[n+1][150];
        for(int sub_array[]:dp)
         Arrays.fill(sub_array,-1);
        return helper_function(s,k,0,'#',dp); // taken # as dummy char becuase in ascii table # is adjcent of z , so it will automatically cover by 150 dimension of dp
        
    }
}

File: 2370-longest-ideal-subsequence/2370-longest-ideal-subsequence.cpp
class Solution {
    int dp[100005][28];
int longest(string &s , int last , int n , int k){
    if(n == 0)
        return  dp[n][last]= 0;
    if(dp[n][last] != -1)
        return dp[n][last];
    else{
        if(last == 27)
            dp[n][last] =  max(1+longest(s,(s[n-1] - 'a'),n-1,k) , longest(s,last,n-1,k));
        else if( abs(last - (s[n-1] - 'a')) > k )
            dp[n][last] = longest(s,last,n-1,k);
        else
            dp[n][last] = max(1 + longest(s,s[n-1]-'a',n-1,k) , longest(s,last,n-1,k));
    }
    
    return dp[n][last];
}
    public:
    int longestIdealString(string s, int k) {
        int n = s.size();
        memset(dp,-1,sizeof(dp));
        return longest(s,27,n,k);
    }
};